determined-ai/determined
# fix: replace tree with code mirror for metadata view

## Ticket




## Description

Replace metadata tree view with code mirror to show metadata. 

### Reason

The tree view does not handle all edge cases to show metadata such as nested array, objects in array etc. It would be difficult to handle all edge cases and maintain it, so proposed to use [Code Mirror Editor](https://uiwjs.github.io/react-codemirror/) to show metadata ([Slack thread](https://hpe-aiatscale.slack.com/archives/C06G9QEKUJZ/p1724269555897989)). Especially we already use it in `Code` tab in the experiment/run detail page. Code mirror editor has features to expand/collapse code block and search metadata with regex. Also, it's able to show any complex or big metadata easily, and it's very easy to maintain it because of out of box features. That's why we chose code mirror editor.
Other option was to use JSON viewer, but that means we need to add another dependency which we didn't prefer.


#### Bug in the tree view (Before this PR)

- not correctly show metadata
- every time i click a expand button, metadata is getting bigger and bigger (see the scroll bar in the video)

https://github.com/user-attachments/assets/12ba8b9a-0145-4d95-971d-022912dd54f9


### After

https://github.com/user-attachments/assets/abec8323-0eaf-4f36-991e-5fcb8520c199


## Test Plan

- add complex metadata such as nested array or objects in array etc for a run
- go to the run detail page
- Check if metadata is visible at the bottom
- Check if code blocks are expandable/collapsable
- Check if theres no regressions in Code tab in experiment/run detail page as well
  - code highlight


## Checklist

  See [Release Note](https://github.com/determined-ai/determined/blob/master/docs/release-notes/README.md) for details.




[DET-123]: https://hpe-aiatscale.atlassian.net/browse/DET-123?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ
